### PR TITLE
Image alternate text

### LINKED ISSUE #24: Image alternate text
### UI/UX (taken from [here](https://github.com/ckeditor/ckeditor5-image/issues/16))
1. Adding alternate text icon to image toolbar:
image
2. After pressing it, image toolbar is closed and alternate text balloon is showed:
image
("Save" should be changed to "Ok")
3. After pressing either "Ok" or "Cancel" we go back to 1.

### Implementation
Needs to be implemented:
* command changing alt attribute of selected image, `ImageAlternateTextCommand`, registered in editor during initialization of `ImageEngine`,
* alternate text button that can be added to image toolbar, registered in editor's component factory,
* UI form showed after button is pressed.

### LINKED ISSUE #5043: Image alternative text
### UI/UX (taken from [here](https://github.com/ckeditor/ckeditor5-image/issues/16))
1. Adding alternate text icon to image toolbar:
image
2. After pressing it, image toolbar is closed and alternate text balloon is showed:
image
("Save" should be changed to "Ok")
3. After pressing either "Ok" or "Cancel" we go back to 1.

### Implementation
Needs to be implemented:
* command changing alt attribute of selected image, `ImageAlternateTextCommand`, registered in editor during initialization of `ImageEngine`,
* alternate text button that can be added to image toolbar, registered in editor's component factory,
* UI form showed after button is pressed.

### PR DESCRIPTION
**Pull request by szymonkups**
Fixes ckeditor/ckeditor5#5043 .
I've extracted `ui/imageballonpanel` that is used by `ImageToolbar` and `ImageAlternateText`. This will propably change when we will introduce more general solution for balloon panels inside the editor: https://github.com/ckeditor/ckeditor5-ui/issues/134.

Some issues:
 * Manual test shows single button which is not aligned properly: https://github.com/ckeditor/ckeditor5-theme-lark/issues/69
 * When pressing cancel on the alternate text balloon panel and entering it again textarea is not updated, it will be fixed when proper solution will be find: https://github.com/ckeditor/ckeditor5-ui/issues/135

### CODE REVIEW

**Comment by Reinmar** [comment]
First three observations:

1. The alt of an image doesn't change after submitting the form.
2. It makes no sense for me to use a textarea for the alt... I know @oleq proposed that, but it just seems weird for me. I don't think you'd actually type that much.
3. The buttons have wrong labels. Please see the link balloon.

File: src/imagealternatetext/ui/alternatetextformview.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		 * @member {module:ui/labeledinput/labeledinputview~LabeledInputView} #labeledTextarea
+		 */
+		this.labeledTextarea = this._createLabeledTextarea();
+
+		/**
+		 * Button used to submit the form.
+		 *
+		 * @member {module:ui/button/buttonview~ButtonView} #saveButtonView
+		 */
+		this.saveButtonView = this._createButton( 'Ok' );
```
Comment (Reinmar) [inline src/imagealternatetext/ui/alternatetextformview.js]: Use `t()` directly with a string.

File: src/imagealternatetext/ui/alternatetextformview.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			]
+		} );
+
+		submitHandler( {
+			view: this
+		} );
+	}
+
+	/**
+	 * Creates button View.
```
Comment (Reinmar) [inline src/imagealternatetext/ui/alternatetextformview.js]: s/view/View

File: src/imagealternatetext/ui/textareaview.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import View from '@ckeditor/ckeditor5-ui/src/view';
+import Template from '@ckeditor/ckeditor5-ui/src/template';
+import '../../../theme/textareaview/theme.scss';
+
+/**
+ * The TextAreaView class.
+ *
+ * @extends module:ui/view~View
+ */
+export default class TextAreaView extends View {
```
Comment (Reinmar) [inline src/imagealternatetext/ui/textareaview.js]: I think it should rather be `TextareaView`. cc @oleq?

File: src/ui/imageballoonpanel.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+};
+
+/**
+ * Image balloon panel class. It extends {module:ui/balloonpanel/balloonpanelview~BalloonPanelView} by adding helpers
+ * to use with image widgets. It sets proper positioning on `scroll` and `resize` events and hides the panel when
+ * image is no longer selected or focus is lost.
+ *
+ * @extends module:ui/balloonpanel/balloonpanelview~BalloonPanelView
+ */
+export default class ImageBalloonPanel extends BalloonPanelView {
```
Comment (Reinmar) [inline src/ui/imageballoonpanel.js]: Should be `ImageBalloonPanelView`.

File: src/ui/imageballoonpanel.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		 */
+		this._throttledAttach = throttle( () => {
+			this._attach();
+		}, 100 );
+	}
+
+	/**
+	 * Attaches the panel and enables `scroll` and `resize` listeners.
+	 */
+	attach() {
```
Comment (Reinmar) [inline src/ui/imageballoonpanel.js]: I don't get the difference between `_attach()` and `attach()`.

File: src/imagealternatetext/ui/textareaview.js (hunk 3 above)
Comment (Reinmar) [reply src/imagealternatetext/ui/textareaview.js]: BTW, it'd rather need to end up in the ckeditor5-ui. But, as I commented already, an input would make more sense to me.

Comment (oleq) [reply src/imagealternatetext/ui/textareaview.js]: Yeah, let's use an input. We already have `LabeledInputView` for that kind of stuff in ui.

Comment (Reinmar) [reply src/imagealternatetext/ui/textareaview.js]: Good. This will reduce the amount of code.

File: src/ui/imageballoonpanel.js (hunk 5 above)
Comment (szymonkups) [reply src/ui/imageballoonpanel.js]: I've added private `_attach` method to separate actual panel attaching from listeners setup code. I can then easily throttle only `_attach` method and use reference to that method to remove listeners.

**Comment by Reinmar** [comment]
I've reported a couple of issues, e.g. very important https://github.com/ckeditor/ckeditor5-image/issues/37, but overall this is fine. We also must think about this: https://github.com/ckeditor/ckeditor5-ui/issues/134.

**Pull request merged**